### ISSUE TITLE
Usage of combineLatest

### ISSUE THREAD

**Issue by adamski** [martinfinke/ReaX#15]
I have a set of `Observable<int>`s (actually `BehaviourSubject`s) that I want to use in order to calculate an `Observable<bool>`. In this case, I want to know if all values are 0 or not.  If they are, emit `true`, else `false`. 

Each Observable is a member of an object in a `std::array`.

I think I need to use `combineLatest`. I'm getting confused as to how to set this up. 
Something like this:
```cpp
allTracksClear = Observable<bool> ([this] (Observable<int> t0,Observable<int> t1,Observable<int> t2,Observable<int> t3) {
    return t0 == t1 == t2 == t3 == 0;
},
tracks[0]->loopLengthObservable,
tracks[1]->loopLengthObservable,
tracks[2]->loopLengthObservable,
tracks[3]->loopLengthObservable);
```
But that seems overly verbose.

Also, can you point me to some good documentation/tutorials as I feel this is more of a lack of understanding of how ReactiveX works in general.

Thanks

**Comment by adamski** [comment]
Thanks @martinfinke super helpful.

**Comment by adamski** [comment]
Re: removing the `tracks[0]->loopLengthObservable, tracks[1]->loopLengthObservable, ...`, this could be achieved with another helper method, that takes an array/vector and returns each given member as an array... something like `zip` or `map`?

**Comment by martinfinke** [comment]
> this could be achieved with another helper method, that takes an array/vector and returns each given member as an array... something like zip or map?

Yes, exactly. We would need a `map` function for the `tracks` array. In C++ STL, that's `std::transform`:

```cpp
// These would be given as members in some class
constexpr size_t numTracks = 4;
std::array<std::shared_ptr<Track>, numTracks> tracks;

// And then in some local scope:
std::array<Observable<int>, numTracks> loopLengthObservables;
std::transform(tracks.begin(),
               tracks.end(),
               loopLengthObservables.begin(),
               [](std::shared_ptr<Track> track) {
                   return track->loopLengthObservable;
               });

// Now we can continue like above:
allTracksClear = toArrayObservable(loopLengthObservables)
    .map(...)
```

The good thing here: If you increase `numTracks` in a few months, this code will still cover all tracks. Whereas the solutions we earlier would keep compiling but only check the first four tracks.

I don't really like the syntax of `std::transform` with `begin()`, `end()` etc. so an alternative would be like:

```cpp
// For C++17 and later, replace "result_of" with "invoke_result"
template<typename T, size_t N, typename Function>
typename std::result_of<Function(T)>::type mapArray(const std::array<T, N>& array, const Function& function)
{
    std::array<typename std::result_of<Function(T)>::type, N> results;
    std::transform(array.begin(), array.end(), results.begin(), function);
    return results;
}

// Now you can write:
auto loopLengths = mapArray(tracks,
                            [](auto track){ return track->loopLengthObservable; });
allTracksClear = toArrayObservable(loopLengthObservables)
    .map(...)
```

In my opinion, the question with writing functional C++ is often "Sure it's possible, but is it actually more readable?" With the clutter of `typename`, `result_of`, `::type` and so on, I'm not so sure ;) And the compiler errors aren't more readable, either..

**Comment by adamski** [comment]
Eek! That does start to get complicated... I see what you mean. I'm sure there is a balance to be had there. Thanks again for the detailed and thoughtful answers.

**Comment by martinfinke** [comment]
You're welcome!
